### ISSUE TITLE
Proposing an extension to @rule syntax (mirror julia syntax)

### ISSUE THREAD

**Issue by peterahrens** [JuliaSymbolics/SymbolicUtils.jl#344]
Wouldn't it be nifty to match segment variables with `~a...`? I think it would be nice. It wouldn't be a difficult feature to add either, since we can detect splats in quoted expressions. The feature is non-breaking, since we can keep the old syntax. Furthermore, we need only make a change to `makepattern`. Something like this:

```
if expr.head === :call
    if expr.args[1] === :(~)
        #proceed normally...
elseif expr.head === :...
    if expr.args[1].args[1] === :(~)
        makesegment(expr.args[1].args[2].args[2], keys)
    end
else
#proceed normally...
```

**Comment by 0x0f0f0f** [comment]
I agree

**Comment by 0x0f0f0f** [comment]
I'm currently porting the SU syntax and rule language to Metatheory.jl, I'd redesign the SU syntax for rules but it's hard to make those breaking changes and we would have to PR all packages downstream. I would like to remove the `~` for pattern variables and use `...` for slot variables only, leaving the predicate syntax as is. Also, I would use `|>` for rules that evaluate the rhs (SU rules) and `=>` for rules that perform symbolic substitution only.
```
(~a*~b)*(~c...)-->~a*(~b*~c...)
```
is not really that readable, i'd propose
```
(a*b)*c...=>a*(b*c...)
```
 @peterahrens

**Comment by peterahrens** [comment]
The `makepattern` macro relies on `~` to determine which symbols to turn into slots. Without it, we would need to declare which variables are slots or escape non-slot symbols. If I had a function `f` and a literal `a`, which of the following syntaxes would you prefer?

`@rule b c c * f(a, b) => g(c, b)`
`@rule c * $f($a, b) => g(c, b)`
`@rule c_ * f(a, b_) => g(c, b)` (This is the MacroTools.jl syntax)

The following option isn't purely syntactic; we would traverse the lhs to discover which nodes are typed as slots. The traversal isn't difficult, but it doesn't work well with the @capture macro (#342) because the @capture macro needs a list of variable names to inject into the environment at macro expansion time.
```
@slots b c
@rule c * f(a, b) => g(c, b)
```

I agree that the `~` syntax is a little messy, but I'm not sure the other options are better. I wish that `~` bound tighter. I was hoping to update my proposal for the capture macro with a better syntax "once I figured it out", but that didn't happen.

If we do discover a syntax that we really like, we might avoid backporting by making a separate macro (`@rule2`) that we use instead?

**Comment by peterahrens** [comment]
A fifth option, which I feel is somewhat promising:
`@slots b c @rule c * f(a, b) => g(c, b)`
or the multi-line form
```
@slots b c begin
rw = Postwalk(Chain([
    @rule c * f(a, b) => g(c, b)
    @rule b * c => c * b
]))
end
```

The way this would work: the `@slots` macro would traverse the inner expression and expand each @rule invocation with the slots `b` and `c`, so that `@rule c * f(a, b) => g(c, b)` would be expanded into `@rule b c c * f(a, b) => g(c, b)` which we could evaluate.

**Comment by peterahrens** [comment]
Regardless of how we redesign the parsing of the lhs, I think that it would make the most sense to inject the slot matches as normal variables into the scope of the rhs, so we wouldn't really need any special syntax processing for the rhs.

**Comment by 0x0f0f0f** [comment]
Check my latest Metatheory.jl pull request in this org. Im downporting everything that can be made generic with TermInterface.jl and unifying into a single, general purpose term rewriting library. I have made exactly this change you are proposing in the rule macro in src/Rules/rule_dsl.jl

**Comment by 0x0f0f0f** [comment]
More than that, there are cases if we want this solution to be ported to MT where we want to match both against function objects and symbol literals. The current Metatheory.jl syntax for rules (see tests subdirs like cas/cas_theory in master branch) currently provides four macros `@theory` and `@methodtheory` (same goes for `@rule`) where the first matches against function name symbols  and the latter resolves the symbols and interpolates the function object reference in the pattern. This way, metatheory.jl syntax was roughly this:
`@rule a * f(b, $some_literal) => b`
With no need to interpolate `f` or `*`. This could be easily solved by adding a short-circuiting `&&` in the pattern matcher when checking if the function head matches. First you try the symbol then you resolve the object if you have no luck. Rule objects would then have to store a module pointer to where they were defined in order to solve the `GlobalRef`, but I think this approach would provide a better syntax without manipulation of the RHS and interoperability of matches between multiple contexts. In Symbolics.jl we may want to match against function objects, but when rewriting `Expr` we eant to match against symbols or both.

**Comment by 0x0f0f0f** [comment]
> Check my latest Metatheory.jl pull request in this org. Im downporting everything that can be made generic with TermInterface.jl and unifying into a single, general purpose term rewriting library. I have made exactly this change you are proposing in the rule macro in src/Rules/rule_dsl.jl

Meaning that the RHS functions are built by accepting the pattern variables as arguments, so that you do not need to do manipulation of the RHS expr if you use the 'elegant' ~-less syntax ;)

**Comment by 0x0f0f0f** [comment]
Vexata quaestio @shashi

**Comment by peterahrens** [comment]
I proposed a lot of changes just now. What change are you referring to?

**Comment by peterahrens** [comment]
When you say "I have made exactly this change you are proposing," after reading https://github.com/JuliaSymbolics/Metatheory.jl/pull/77, I'm unclear on what change you are referring to. When you say that you want to "match both against function objects and symbol literals", does that mean that you'd like slots to be able to be used as both operators and arguments?

**Comment by 0x0f0f0f** [comment]
> When you say "I have made exactly this change you are proposing," after reading [JuliaSymbolics/Metatheory.jl#77](https://github.com/JuliaSymbolics/Metatheory.jl/pull/77), I'm unclear on what change you are referring to. When you say that you want to "match both against function objects and symbol literals", does that mean that you'd like slots to be able to be used as both operators and arguments?

Sorry, I forgot to quote reply. I meant the manipulation of the RHS can be avoided if we pass the pattern variables as function arguments to the rhs

**Comment by 0x0f0f0f** [comment]
> When you say "I have made exactly this change you are proposing," after reading [JuliaSymbolics/Metatheory.jl#77](https://github.com/JuliaSymbolics/Metatheory.jl/pull/77), I'm unclear on what change you are referring to. When you say that you want to "match both against function objects and symbol literals", does that mean that you'd like slots to be able to be used as both operators and arguments?

Nope. I mean this (sorry if i was unclear)
`f(a,b) => c` should (or could?) match against both `:(f(1,2))` and `:(($f)(1,2))`

**Comment by peterahrens** [comment]
Ah, I see. So setting `f = :f` should work, although perhaps a different macro mode to quote symbols would be more convenient for you. And it sounds like when you say "pass the pattern variables as function arguments to the rhs," you mean that `(@rule a => println(a))("foo")` should either
1. print `Slot(:a),` and evaluate to nothing
2. print nothing and evaluate to :(println("foo"))
3. print nothing and evaluate to :(println(Slot(:a)))
whereas I was hoping it would
4. print `"foo"` and evaluate to nothing.

**Comment by peterahrens** [comment]
I think that I'm starting to understand. It sounds like the goal for @0x0f0f0f is to use the same syntax in SymbolicUtils and MetaTheory.jl, where the syntax of SymbolicUtils `@rule` is a subset of the syntax available in Metatheory, is that correct?

If that is the case, I think the discussion regarding syntax for
1. whether matching results are interpolated in the rhs
2. whether or not the rhs gets evaluated
3. whether non-slot symbols are quoted or evaluated in the lhs
belongs in https://github.com/JuliaSymbolics/Metatheory.jl/issues/68.

In this thread, we need only to propose
A. How to declare which variables are slots
B. How to declare which of those are segments (I believe our proposal is to use `...`)

Once there's a proposal for A, I would be happy to draft up a PR. I think that A and B are orthogonal decisions from 1, 2, or 3.

**Comment by 0x0f0f0f** [comment]
> 4. print `"foo"` and evaluate to nothing, the current behavior of the `@rule` macro in SymbolicUtils.jl

**Comment by 0x0f0f0f** [comment]
I'll make a module `Rules.NewSyntax` in MT to re-introduce the metatheory syntax so we can compare them, I'll now default to SU's current syntax in my MT PR. @peterahrens

**Comment by peterahrens** [comment]
I'm slowly gravitating towards the following syntax, because it results in rules that look clean, and it's backwards compatible. You can still use ~ to declare slots and ~~ to declare segments, if you'd like. The only difference is that the `@rule` macro would now take more than one argument. Any thoughts on this syntax?

`@slots b c @rule c * f(a, b) => g(c, b)`
`@rule b c c * f(a, b) => g(c, b)`
```
@slots b c begin
rw = Postwalk(Chain([
    @rule c * f(a, b) => g(c, b)
    @rule b * c... => *(c..., b)
]))
end
```

**Comment by 0x0f0f0f** [comment]
Implemented in #368
